src/VisualStudio/IntegrationTests/AbstractEditorTests.cs

```diff
@@ -0,0 +1,231 @@
+
+                if (trimWhitespace)
+                {
+                    if (caretStartIndex == 0)
+                    {
+                        lineText = lineText.TrimEnd();
+                    }
+                    else if (caretEndIndex == expectedText.Length)
+                    {
+                        lineText = lineText.TrimStart();
+                    }
+                    else
+                    {
+                        lineText = lineText.Trim();
+                    }
+                }
+
+                var lineTextBeforeCaret = caretStartIndex < lineText.Length
+                    ? lineText.Substring(0, caretStartIndex)
+                    : lineText;
+
+                var lineTextAfterCaret = caretStartIndex < lineText.Length
+                    ? lineText.Substring(caretStartIndex)
+                    : string.Empty;
+
+                Assert.Equal(expectedTextBeforeCaret, lineTextBeforeCaret);
+                Assert.Equal(expectedTextAfterCaret, lineTextAfterCaret);
+                Assert.Equal(expectedTextBeforeCaret.Length + expectedTextAfterCaret.Length, lineText.Length);
+            }
+            else
+            {
+                var lineText = _editor.GetCurrentLineText();
+
+                if (trimWhitespace)
+                {
+                    lineText = lineText.Trim();
+                }
+
+                Assert.Equal(expectedText, lineText);
+            }
+        }
+
+        protected void VerifyTextContains(string expectedText)
+        {
+            var caretStartIndex = expectedText.IndexOf("$$");
+
+            if (caretStartIndex >= 0)
+            {
+                var caretEndIndex = caretStartIndex + "$$".Length;
+
+                var expectedTextBeforeCaret = caretStartIndex < expectedText.Length
+                    ? expectedText.Substring(0, caretStartIndex)
+                    : expectedText;
+
+                var expectedTextAfterCaret = caretEndIndex < expectedText.Length
+                    ? expectedText.Substring(caretEndIndex)
+                    : string.Empty;
+
+                var expectedTextWithoutCaret = expectedTextBeforeCaret + expectedTextAfterCaret;
+
+                var editorText = _editor.GetText();
+                Assert.Contains(expectedTextWithoutCaret, editorText);
+
+                var index = editorText.IndexOf(expectedTextWithoutCaret);
+
+                var caretPosition = _editor.GetCaretPosition();
+                Assert.Equal(caretStartIndex + index, caretPosition);
+            }
+            else
+            {
+                var editorText = _editor.GetText();
+                Assert.Contains(expectedText, editorText);
+            }
+        }
+
+        protected void VerifyCompletionItemExists(params string[] expectedItems)
+        {
+            WaitForAsyncOperations(FeatureAttribute.CompletionSet);
+
+            var completionItems = _editor.GetCompletionItems();
```

Comment:
We you able to figure out any elegant way of making GetCompletionItems implicitly do the WaitForAsyncOperations()? There's some layering challenges, but enough meditation over design patterns might help?